### PR TITLE
Add baremetal section with power-cycle remediation to machine health checks doc 

### PR DESCRIPTION
**Pull request by n1r1**
Modify machine health checks documentation to include an option to trigger power-cycle instead of machine deletion.
This is only relevant for IPI baremetal clusters and starting from OCP 4.5

### CODE REVIEW

**Comment by n1r1** [comment]
cc @JoelSpeed @beekhof

**Comment by vikram-redhat** [comment]
@kalexand-rh can you take a look? @n1r1 which version does it apply to?

**Comment by n1r1** [comment]
@vikram-redhat  
4.5 + (i.e 4.5, 4.6)

File: modules/machine-health-checks-baremetal.adoc [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// Module included in the following assemblies:
+//
+// * machine_management/deploying-machine-health-checks.adoc
+// * post_installation_configuration/node-tasks.adoc
+
+[id="machine-health-checks-baremetal_{context}"]
```
Comment (kalexand-rh) [inline modules/machine-health-checks-baremetal.adoc:6]: You don't need to add another module. You need to weave this content into the existing modules.

File: modules/machine-health-checks-baremetal.adoc [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+= MachineHealthChecks on Bare Metal
+
+Machine deletion on bare metal cluster triggers reprovisioning of a bare metal host.
```
Comment (kalexand-rh) [inline modules/machine-health-checks-baremetal.adoc:9]: I think the bulk of the content needs to go in modules/machine-health-checks-about.adoc around line 30.

File: modules/machine-health-checks-baremetal.adoc [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+Usually bare metal reprovisioning is a lengthy process, in which the cluster
```
Comment (JoelSpeed) [inline modules/machine-health-checks-baremetal.adoc:10]: I had to read this a couple of times to understand it, `during` may be more obvious what you mean here
```suggestion
Usually bare metal reprovisioning is a lengthy process, during which the cluster 
```

File: modules/machine-health-checks-baremetal.adoc [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+//
+// * machine_management/deploying-machine-health-checks.adoc
+// * post_installation_configuration/node-tasks.adoc
+
+[id="machine-health-checks-baremetal_{context}"]
+= MachineHealthChecks on Bare Metal
+
+Machine deletion on bare metal cluster triggers reprovisioning of a bare metal host.
+Usually bare metal reprovisioning is a lengthy process, in which the cluster 
+is lacking compute power, and some applications are down.
```
Comment (JoelSpeed) [inline modules/machine-health-checks-baremetal.adoc:11]: Not a guarantee that losing one host would take down applications so maybe reword as below, WDYT?
```suggestion
is missing compute resources and applications may be interrupted.
```

**Review by JoelSpeed — COMMENTED** [review]
Added a couple of suggestions to improve the wording, but otherwise I think this is good

File: modules/machine-health-checks-baremetal.adoc [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+To change the default remediation process from machine deletion to host power-cycle,
+the MachineHealthCheck resource should be annotated with
```
Comment (kalexand-rh) [inline modules/machine-health-checks-baremetal.adoc:13]:
```suggestion
annotate the MachineHealthCheck resource with the
```

File: modules/machine-health-checks-baremetal.adoc [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+[id="machine-health-checks-baremetal_{context}"]
+= MachineHealthChecks on Bare Metal
+
+Machine deletion on bare metal cluster triggers reprovisioning of a bare metal host.
+Usually bare metal reprovisioning is a lengthy process, in which the cluster 
+is lacking compute power, and some applications are down.
+To change the default remediation process from machine deletion to host power-cycle,
+the MachineHealthCheck resource should be annotated with 
+`machine.openshift.io/remediation-strategy: external-baremetal`
```
Comment (kalexand-rh) [inline modules/machine-health-checks-baremetal.adoc:14]:
```suggestion
`machine.openshift.io/remediation-strategy: external-baremetal` annotation.
```

File: modules/machine-health-checks-baremetal.adoc [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+Once the annotation is set, unhealthy machines will be power-cycled by using
```
Comment (kalexand-rh) [inline modules/machine-health-checks-baremetal.adoc:16]:
```suggestion
After you set the annotation set, unhealthy machines are power-cycled by using 
```

File: modules/machine-health-checks-baremetal.adoc (hunk 6 above)
Comment (kalexand-rh) [inline modules/machine-health-checks-baremetal.adoc:14]: You also need to add this parameter to modules/machine-health-checks-resource.adoc and add an annotation that explains what it is, like:

```
<37> For bare metal clusters, you must add this annotation. Do not add this parameter to any other cluster type.
```

**Review by kalexand-rh — COMMENTED** [review]
I think this is a great addition, but I think the information would be more effective in different modules.

File: modules/machine-health-checks-baremetal.adoc (hunk 4 above)
Comment (kalexand-rh) [reply modules/machine-health-checks-baremetal.adoc:11]: s/may/might

File: modules/machine-health-checks-baremetal.adoc (hunk 6 above)
Comment (n1r1) [reply modules/machine-health-checks-baremetal.adoc:14]: I'm concerned that some users will copy the yaml as is and use it, even on non-baremetal cluster which can end up with remediation not working.
WDYT from your experience in other areas?

**Comment by n1r1** [comment]
Thank you both for the feedback.
I made the changes per your suggestions.

**Comment by JoelSpeed** [comment]
LGTM

File: modules/machine-health-checks-about.adoc [hunk 8]
```diff
@@ -29,3 +29,16 @@ targeted pool of machines, remediation stops so that manual intervention can tak
 place.
 
 To stop the check, you remove the resource.
+
+= MachineHealthChecks on Bare Metal
```
Comment (kalexand-rh) [inline modules/machine-health-checks-about.adoc:33]:
```suggestion
[id="machine-health-checks-bare-metal_{context}"]
== MachineHealthChecks on Bare Metal
```

File: modules/machine-health-checks-about.adoc [hunk 9]
```diff
@@ -29,3 +29,16 @@ targeted pool of machines, remediation stops so that manual intervention can tak
...
+
+Machine deletion on bare metal cluster triggers reprovisioning of a bare metal host.
+Usually bare metal reprovisioning is a lengthy process, during which the cluster
+is missing compute resources, and applications might be interrupted.
```
Comment (kalexand-rh) [inline modules/machine-health-checks-about.adoc:37]:
```suggestion
is missing compute resources and applications might be interrupted.
```

File: modules/machine-health-checks-baremetal.adoc (hunk 6 above)
Comment (kalexand-rh) [reply modules/machine-health-checks-baremetal.adoc]: Is there a value for this annotation that's valid for non-baremetal clusters? If so, I'd put that in the YAML and explain the available options, including this one, in the annotation. I expect that people are going to mess this up otherwise because the information is fairly far from where you use it.

Comment (n1r1) [reply modules/machine-health-checks-baremetal.adoc]: > Is there a value for this annotation that's valid for non-baremetal clusters?

Unfortunately not. This annotation is only for bare metal with only one valid value.
I'll be happy to do whatever you think is best for the users

Comment (kalexand-rh) [reply modules/machine-health-checks-baremetal.adoc]: Then add a callout to the stanza that would include that annotation and provide the entire annotation in the callout text.

Comment (n1r1) [reply modules/machine-health-checks-baremetal.adoc]: @kalexand-rh makes sense.
I hope I got it right.

File: modules/machine-health-checks-resource.adoc [hunk 10]
```diff
@@ -17,27 +17,30 @@ kind: MachineHealthCheck
 metadata:
   name: example <1>
   namespace: openshift-machine-api
+  annotations:
+     <2>
```
Comment (kalexand-rh) [inline modules/machine-health-checks-resource.adoc:21]:
```suggestion

```

File: modules/machine-health-checks-resource.adoc [hunk 11]
```diff
@@ -17,27 +17,30 @@ kind: MachineHealthCheck
...
 spec:
   selector:
     matchLabels:
```
Comment (kalexand-rh) [inline modules/machine-health-checks-resource.adoc:24]:
```suggestion
    matchLabels: <2>
```

File: modules/machine-health-checks-resource.adoc [hunk 12]
```diff
@@ -17,27 +17,30 @@ kind: MachineHealthCheck
...
-    timeout: "300s" <4>
+    timeout: "300s" <5>
     status: "Unknown"
-  maxUnhealthy: "40%" <5>
+  maxUnhealthy: "40%" <6>
 ----
 <1> Specify the name of the MachineHealthCheck to deploy.
-<2> Specify a label for the machine pool that you want to check.
-<3> Specify the MachineSet to track in `<cluster_name>-<label>-<zone>`
+<2> Add `machine.openshift.io/remediation-strategy: external-baremetal` annotation to enable power-cycle remediation for bare metal.
```
Comment (kalexand-rh) [inline modules/machine-health-checks-resource.adoc:38]:
```suggestion
<2> For bare metal clusters, add the `machine.openshift.io/remediation-strategy: external-baremetal` annotation to the `matchLabels` section to enable power-cycle remediation. With this remediation strategy, unhealthy hosts are rebooted instead of removed from the cluster.
```

**Review by kalexand-rh — COMMENTED** [review]
@n1r1, thank you! I have a few more suggestions, but this is really good.

**Comment by n1r1** [comment]
Thanks @kalexand-rh 
I'm not sure I understand the last suggestions.

The annotation is not related to the `matchLabel` sections.
Here's a full example with that annotation, if it helps to clarify things:
```
apiVersion: machine.openshift.io/v1beta1
kind: MachineHealthCheck
metadata:
  name: example 
  namespace: openshift-machine-api
  annotations:
    machine.openshift.io/remediation-strategy: external-baremetal
spec:
  selector:
    matchLabels:
      machine.openshift.io/cluster-api-machine-role: <role> 
      machine.openshift.io/cluster-api-machine-type: <role> 
      machine.openshift.io/cluster-api-machineset: <cluster_name>-<label>-<zone> 
  unhealthyConditions:
  - type:    "Ready"
    timeout: "300s" 
    status: "False"
  - type:    "Ready"
    timeout: "300s" 
    status: "Unknown"
  maxUnhealthy: "40%" 
```

File: modules/machine-health-checks-resource.adoc (hunk 12 above)
Comment (n1r1) [reply modules/machine-health-checks-resource.adoc:39]: Thanks @kalexand-rh 
I'm not sure I understand the last suggestions.

The annotation is not related to the `matchLabel` sections.
Here's a full example with that annotation, if it helps to clarify things:
```
apiVersion: machine.openshift.io/v1beta1
kind: MachineHealthCheck
metadata:
  name: example 
  namespace: openshift-machine-api
  annotations:
    machine.openshift.io/remediation-strategy: external-baremetal
spec:
  selector:
    matchLabels:
      machine.openshift.io/cluster-api-machine-role: <role> 
      machine.openshift.io/cluster-api-machine-type: <role> 
      machine.openshift.io/cluster-api-machineset: <cluster_name>-<label>-<zone> 
  unhealthyConditions:
  - type:    "Ready"
    timeout: "300s" 
    status: "False"
  - type:    "Ready"
    timeout: "300s" 
    status: "Unknown"
  maxUnhealthy: "40%" 
```

Comment (kalexand-rh) [reply modules/machine-health-checks-resource.adoc:39]: @n1r1, thanks for the clarification. I've updated the suggestion.

Comment (n1r1) [reply modules/machine-health-checks-resource.adoc]: @kalexand-rh , thanks for the suggestion, it makes sense and I've now accepted it.
Please let me know if anything else is needed

File: modules/machine-health-checks-resource.adoc (hunk 10 above)
Comment (n1r1) [reply modules/machine-health-checks-resource.adoc:21]: @kalexand-rh is this suggestion still relevant per our last discussion?

If I'll remove this, we will have `<2>` below explaining about the annotation without any reference in the yaml. 
let me know if you want to remove it anyway

Comment (kalexand-rh) [reply modules/machine-health-checks-resource.adoc]: @n1r1, I think it might be clearest to provide two yaml sections.  Will you PTAL at the commit that I've added?

Comment (n1r1) [reply modules/machine-health-checks-resource.adoc]: @kalexand-rh thanks.
Looks good to me
